### PR TITLE
fix: handle optimism deposit transactions on `SenderRecovery` stage

### PR DESCRIPTION
**Pull request by joshieDo**
Deposit transactions do not have a signature, and we were trying to recover them during the stage.

### CODE REVIEW

File: crates/primitives/src/transaction/mod.rs [hunk 1]
```diff
@@ -856,10 +856,37 @@ impl TransactionSignedNoHash {
...
     }
 
+    /// Recover signer from signature and hash _without ensuring that the signature has a low `s`
+    /// value_. 
+    /// 
+    /// Re-uses a given buffer to avoid numerous reallocations when recovering batches.
+    ///
+    /// Returns `None` if the transaction's signature is invalid, see also
+    /// [Signature::recover_signer_unchecked].
+    pub fn recover_signer_unchecked_with_buffer(&self, rlp_buf: &mut Vec<u8>) -> Option<Address> {
```
Comment (joshieDo) [inline crates/primitives/src/transaction/mod.rs:877]: this is a mouthful, accepting suggestions

File: crates/primitives/src/transaction/mod.rs [hunk 2]
```diff
@@ -856,10 +856,37 @@ impl TransactionSignedNoHash {
...
     }
 
+    /// Recover signer from signature and hash _without ensuring that the signature has a low `s`
+    /// value_.
+    ///
+    /// Re-uses a given buffer to avoid numerous reallocations when recovering batches.
+    ///
+    /// Returns `None` if the transaction's signature is invalid, see also
+    /// [Signature::recover_signer_unchecked].
+    pub fn recover_signer_unchecked_with_buffer(&self, rlp_buf: &mut Vec<u8>) -> Option<Address> {
```
Comment (mattsse) [inline crates/primitives/src/transaction/mod.rs:877]: should this always clear the buffer before returning or clear at the beginning?

imo clear first would be safer

File: crates/primitives/src/transaction/mod.rs [hunk 3]
```diff
@@ -856,10 +856,37 @@ impl TransactionSignedNoHash {
...
+        }
+
         let signature_hash = self.signature_hash();
         self.signature.recover_signer(signature_hash)
     }
 
+    /// Recover signer from signature and hash _without ensuring that the signature has a low `s`
+    /// value_.
+    ///
+    /// Re-uses a given buffer to avoid numerous reallocations when recovering batches.
```
Comment (mattsse) [inline crates/primitives/src/transaction/mod.rs:873]: depending on where we clear the buffer, we need to document this

File: crates/primitives/src/transaction/mod.rs (hunk 2 above)
Comment (mattsse) [inline crates/primitives/src/transaction/mod.rs:877]: this is basically, encode + recover unchecked?

so perhaps `encode_recover_unchecked`?

wdyt @Rjected

**Review by mattsse — CHANGES_REQUESTED** [review]
this makes sense,

need to properly document the expections wrt buffer content, perhaps ideally only use the [len..] slice when hashing

**Review by emhane — APPROVED** [review]
Screenshot 2024-03-28 at 16 35 54

pulled this fix into https://github.com/paradigmxyz/reth/pull/6151. defo works as metrics shows, sender recovery stage passed!

File: crates/primitives/src/transaction/mod.rs (hunk 2 above)
Comment (Rjected) [reply crates/primitives/src/transaction/mod.rs]: yeah, that or `encode_and_recover_unchecked` seem fine

**Review by mattsse — CHANGES_REQUESTED** [review]

**Review by mattsse — APPROVED** [review]

**Pull request merged**